File: scripts/contrast_check.py
import math

def hex_to_rgb(hex_color):
    hex_color = hex_color.lstrip('#')
    return tuple(int(hex_color[i:i+2], 16) for i in (0, 2, 4))

def luminance(rgb):
    a = [x / 255.0 for x in rgb]
    a = [((x + 0.055) / 1.055) ** 2.4 if x > 0.03928 else x / 12.92 for x in a]
    return 0.2126 * a[0] + 0.7152 * a[1] + 0.0722 * a[2]

def contrast_ratio(rgb1, rgb2):
    lum1 = luminance(rgb1)
    lum2 = luminance(rgb2)
    brightest = max(lum1, lum2)
    darkest = min(lum1, lum2)
    return (brightest + 0.05) / (darkest + 0.05)

def blend_color(fg_rgb, bg_rgb, alpha):
    return tuple(int(fg * alpha + bg * (1 - alpha)) for fg, bg in zip(fg_rgb, bg_rgb))

# Tailwind Colors (Slate, Teal, Rose, Indigo)
colors = {
    'white': '#ffffff',
    'black': '#000000',
    'slate-50': '#f8fafc',
    'slate-100': '#f1f5f9',
    'slate-200': '#e2e8f0',
    'slate-300': '#cbd5e1',
    'slate-400': '#94a3b8',
    'slate-600': '#475569',
    'slate-700': '#334155',
    'slate-800': '#1e293b',
    'slate-900': '#0f172a',
    'teal-100': '#ccfbf1',
    'teal-400': '#2dd4bf',
    'teal-600': '#0d9488',
    'teal-900': '#134e4a',
    'rose-100': '#ffe4e6',
    'rose-400': '#fb7185',
    'rose-600': '#e11d48',
    'rose-900': '#881337',
    'indigo-100': '#e0e7ff',
    'indigo-400': '#818cf8',
    'indigo-600': '#4f46e5',
    'indigo-900': '#312e81',
}

# Define scenarios
scenarios = [
    # Light Mode
    {'mode': 'Light', 'element': 'Page Background', 'bg': 'slate-50', 'fg': 'slate-900', 'text': 'H1 (About QRCraftly)', 'size': 'large'},
    {'mode': 'Light', 'element': 'Page Background', 'bg': 'slate-50', 'fg': 'slate-600', 'text': 'Intro Paragraph', 'size': 'normal'},
    {'mode': 'Light', 'element': 'Card', 'bg': 'white', 'fg': 'slate-900', 'text': 'Card H3', 'size': 'normal'},
    {'mode': 'Light', 'element': 'Card', 'bg': 'white', 'fg': 'slate-600', 'text': 'Card Paragraph', 'size': 'normal'},
    {'mode': 'Light', 'element': 'Icon Teal', 'bg': 'teal-100', 'fg': 'teal-600', 'text': 'Shield Icon', 'size': 'large'}, # Icons are graphic but good to check
    {'mode': 'Light', 'element': 'Icon Rose', 'bg': 'rose-100', 'fg': 'rose-600', 'text': 'Database Icon', 'size': 'large'},
    {'mode': 'Light', 'element': 'Icon Indigo', 'bg': 'indigo-100', 'fg': 'indigo-600', 'text': 'Code Icon', 'size': 'large'},
    {'mode': 'Light', 'element': 'License Section', 'bg': 'slate-50', 'fg': 'slate-900', 'text': 'License H2', 'size': 'large'},
    {'mode': 'Light', 'element': 'License Section', 'bg': 'slate-50', 'fg': 'slate-600', 'text': 'License P', 'size': 'normal'},
    {'mode': 'Light', 'element': 'Button', 'bg': 'slate-900', 'fg': 'white', 'text': 'Github Button', 'size': 'normal'},

    # Dark Mode (Assuming bg-slate-900 for page background which is standard for dark mode if missing)
    {'mode': 'Dark', 'element': 'Page Background', 'bg': 'slate-900', 'fg': 'white', 'text': 'H1 (About QRCraftly)', 'size': 'large'},
    {'mode': 'Dark', 'element': 'Page Background', 'bg': 'slate-900', 'fg': 'slate-300', 'text': 'Intro Paragraph', 'size': 'normal'},
    {'mode': 'Dark', 'element': 'Card', 'bg': 'slate-800', 'fg': 'white', 'text': 'Card H3', 'size': 'normal'},
    {'mode': 'Dark', 'element': 'Card', 'bg': 'slate-800', 'fg': 'slate-400', 'text': 'Card Paragraph', 'size': 'normal'},
    {'mode': 'Dark', 'element': 'Icon Teal', 'bg': ('teal-900', 0.3, 'slate-800'), 'fg': 'teal-400', 'text': 'Shield Icon', 'size': 'large'},
    {'mode': 'Dark', 'element': 'Icon Rose', 'bg': ('rose-900', 0.3, 'slate-800'), 'fg': 'rose-400', 'text': 'Database Icon', 'size': 'large'},
    {'mode': 'Dark', 'element': 'Icon Indigo', 'bg': ('indigo-900', 0.3, 'slate-800'), 'fg': 'indigo-400', 'text': 'Code Icon', 'size': 'large'},
    {'mode': 'Dark', 'element': 'License Section', 'bg': ('slate-800', 0.5, 'slate-900'), 'fg': 'white', 'text': 'License H2', 'size': 'large'},
    {'mode': 'Dark', 'element': 'License Section', 'bg': ('slate-800', 0.5, 'slate-900'), 'fg': 'slate-300', 'text': 'License P', 'size': 'normal'},
    {'mode': 'Dark', 'element': 'Button', 'bg': 'white', 'fg': 'slate-900', 'text': 'Github Button', 'size': 'normal'},
]

print(f"{'Mode':<6} | {'Element':<20} | {'Contrast':<8} | {'Pass?':<6} | {'Level':<5} | {'Details'}")
print("-" * 80)

for s in scenarios:
    # Resolve BG
    if isinstance(s['bg'], tuple):
        # Blend
        overlay_color = s['bg'][0]
        opacity = s['bg'][1]
        base_color_name = s['bg'][2]

        overlay_rgb = hex_to_rgb(colors[overlay_color])
        base_rgb = hex_to_rgb(colors[base_color_name])

        bg_rgb = blend_color(overlay_rgb, base_rgb, opacity)
    else:
        bg_rgb = hex_to_rgb(colors[s['bg']])

    # Resolve FG
    fg_rgb = hex_to_rgb(colors[s['fg']])

    ratio = contrast_ratio(fg_rgb, bg_rgb)

    # Check WCAG
    # Normal Text: AA >= 4.5, AAA >= 7
    # Large Text (18pt or 14pt bold): AA >= 3, AAA >= 4.5
    # Icons/UI: AA >= 3

    min_ratio = 4.5
    if s['size'] == 'large':
        min_ratio = 3.0

    passed = ratio >= min_ratio

    # AAA check
    aaa_ratio = 7.0
    if s['size'] == 'large':
        aaa_ratio = 4.5
    aaa_passed = ratio >= aaa_ratio

    level = "Fail"
    if aaa_passed:
        level = "AAA"
    elif passed:
        level = "AA"

    print(f"{s['mode']:<6} | {s['text']:<20} | {ratio:.2f}:1   | {'YES' if passed else 'NO':<6} | {level:<5} | {s['bg']} vs {s['fg']}")

File: scripts/contrast_check.js
// Script to check WCAG contrast ratios for various UI elements in QRCraftly.

function hexToRgb(hexColor) {
  hexColor = hexColor.replace(/^#/, "");
  if (hexColor.length === 3) {
    hexColor = hexColor
      .split("")
      .map((c) => c + c)
      .join("");
  }
  return [
    parseInt(hexColor.slice(0, 2), 16),
    parseInt(hexColor.slice(2, 4), 16),
    parseInt(hexColor.slice(4, 6), 16),
  ];
}

function luminance(rgb) {
  const a = rgb.map((x) => {
    x /= 255;
    return x <= 0.03928 ? x / 12.92 : Math.pow((x + 0.055) / 1.055, 2.4);
  });
  return 0.2126 * a[0] + 0.7152 * a[1] + 0.0722 * a[2];
}

function contrastRatio(rgb1, rgb2) {
  const lum1 = luminance(rgb1);
  const lum2 = luminance(rgb2);
  const brightest = Math.max(lum1, lum2);
  const darkest = Math.min(lum1, lum2);
  return (brightest + 0.05) / (darkest + 0.05);
}

function blendColor(fgRgb, bgRgb, alpha) {
  return fgRgb.map((fg, i) => Math.round(fg * alpha + bgRgb[i] * (1 - alpha)));
}

import fs from "fs";

const colorsJson = JSON.parse(
  fs.readFileSync(new URL("../src/colors.json", import.meta.url), "utf-8"),
);
const colors = colorsJson.ui;
const presets = colorsJson.presets;

const scenarios = [
  // Light Mode
  {
    mode: "Light",
    element: "Page Background",
    bg: "slate-50",
    fg: "slate-900",
    text: "H1 (About QRCraftly)",
    size: "large",
  },
  {
    mode: "Light",
    element: "Page Background",
    bg: "slate-50",
    fg: "slate-600",
    text: "Intro Paragraph",
    size: "normal",
  },
  {
    mode: "Light",
    element: "Card",
    bg: "white",
    fg: "slate-900",
    text: "Card H3",
    size: "normal",
  },
  {
    mode: "Light",
    element: "Card",
    bg: "white",
    fg: "slate-600",
    text: "Card Paragraph",
    size: "normal",
  },
  {
    mode: "Light",
    element: "Icon Teal",
    bg: "teal-100",
    fg: "teal-700",
    text: "Shield Icon",
    size: "large",
  },
  {
    mode: "Light",
    element: "Icon Rose",
    bg: "rose-100",
    fg: "rose-700",
    text: "Database Icon",
    size: "large",
  },
  {
    mode: "Light",
    element: "Icon Indigo",
    bg: "indigo-100",
    fg: "indigo-700",
    text: "Code Icon",
    size: "large",
  },
  {
    mode: "Light",
    element: "License Section",
    bg: "slate-50",
    fg: "slate-900",
    text: "License H2",
    size: "large",
  },
  {
    mode: "Light",
    element: "License Section",
    bg: "slate-50",
    fg: "slate-600",
    text: "License P",
    size: "normal",
  },
  {
    mode: "Light",
    element: "Button",
    bg: "slate-900",
    fg: "white",
    text: "Github Button",
    size: "normal",
  },

  // Dark Mode
  {
    mode: "Dark",
    element: "Page Background",
    bg: "slate-900",
    fg: "white",
    text: "H1 (About QRCraftly)",
    size: "large",
  },
  {
    mode: "Dark",
    element: "Page Background",
    bg: "slate-900",
    fg: "slate-300",
    text: "Intro Paragraph",
    size: "normal",
  },
  {
    mode: "Dark",
    element: "Card",
    bg: "slate-800",
    fg: "white",
    text: "Card H3",
    size: "normal",
  },
  {
    mode: "Dark",
    element: "Card",
    bg: "slate-800",
    fg: "slate-400",
    text: "Card Paragraph",
    size: "normal",
  },
  {
    mode: "Dark",
    element: "Icon Teal",
    bg: ["teal-900", 0.3, "slate-800"],
    fg: "teal-400",
    text: "Shield Icon",
    size: "large",
  },
  {
    mode: "Dark",
    element: "Icon Rose",
    bg: ["rose-900", 0.3, "slate-800"],
    fg: "rose-400",
    text: "Database Icon",
    size: "large",
  },
  {
    mode: "Dark",
    element: "Icon Indigo",
    bg: ["indigo-900", 0.3, "slate-800"],
    fg: "indigo-400",
    text: "Code Icon",
    size: "large",
  },
  {
    mode: "Dark",
    element: "License Section",
    bg: ["slate-800", 0.5, "slate-900"],
    fg: "white",
    text: "License H2",
    size: "large",
  },
  {
    mode: "Dark",
    element: "License Section",
    bg: ["slate-800", 0.5, "slate-900"],
    fg: "slate-300",
    text: "License P",
    size: "normal",
  },
  {
    mode: "Dark",
    element: "Button",
    bg: "white",
    fg: "slate-900",
    text: "Github Button",
    size: "normal",
  },
];

function runCheck() {
  console.log(
    `${"Mode".padEnd(6)} | ${"Element".padEnd(20)} | ${"Contrast".padEnd(8)} | ${"Pass?".padEnd(6)} | ${"Level".padEnd(5)} | Details`,
  );
  console.log("-".repeat(80));

  let allPassed = true;

  for (const s of scenarios) {
    let bgRgb;
    if (Array.isArray(s.bg)) {
      const overlayColor = s.bg[0];
      const opacity = s.bg[1];
      const baseColorName = s.bg[2];

      const overlayRgb = hexToRgb(colors[overlayColor]);
      const baseRgb = hexToRgb(colors[baseColorName]);

      bgRgb = blendColor(overlayRgb, baseRgb, opacity);
    } else {
      bgRgb = hexToRgb(colors[s.bg]);
    }

    const fgRgb = hexToRgb(colors[s.fg]);

    const ratio = contrastRatio(fgRgb, bgRgb);

    let minRatio = 4.5;
    if (s.size === "large") {
      minRatio = 3.0;
    }

    const passed = ratio >= minRatio;
    if (!passed) allPassed = false;

    let aaaRatio = 7.0;
    if (s.size === "large") {
      aaaRatio = 4.5;
    }
    const aaaPassed = ratio >= aaaRatio;

    let level = "Fail";
    if (aaaPassed) {
      level = "AAA";
    } else if (passed) {
      level = "AA";
    }

    const bgName = Array.isArray(s.bg) ? `[${s.bg.join(", ")}]` : s.bg;
    console.log(
      `${s.mode.padEnd(6)} | ${s.text.padEnd(20)} | ${ratio.toFixed(2)}:1   | ${(passed ? "YES" : "NO").padEnd(6)} | ${level.padEnd(5)} | ${bgName} vs ${s.fg}`,
    );
  }

  console.log("\nChecking Presets:");
  console.log(
    `${"Preset".padEnd(12)} | ${"Element".padEnd(8)} | ${"Contrast".padEnd(8)} | ${"Pass?".padEnd(6)} | Details`,
  );
  console.log("-".repeat(60));

  for (const p of presets) {
    const bgRgb = hexToRgb(p.bg);
    const fgRgb = hexToRgb(p.fg);
    const eyeRgb = hexToRgb(p.eye);

    const fgRatio = contrastRatio(fgRgb, bgRgb);
    const eyeRatio = contrastRatio(eyeRgb, bgRgb);

    const minRatio = 3.0; // Graphic elements generally require 3.0:1

    const fgPassed = fgRatio >= minRatio;
    const eyePassed = eyeRatio >= minRatio;

    if (!fgPassed || !eyePassed) allPassed = false;

    console.log(
      `${p.label.padEnd(12)} | ${"FG vs BG".padEnd(8)} | ${fgRatio.toFixed(2)}:1   | ${(fgPassed ? "YES" : "NO").padEnd(6)} | ${p.bg} vs ${p.fg}`,
    );
    console.log(
      `${p.label.padEnd(12)} | ${"Eye vs BG".padEnd(8)} | ${eyeRatio.toFixed(2)}:1   | ${(eyePassed ? "YES" : "NO").padEnd(6)} | ${p.bg} vs ${p.eye}`,
    );
  }

  if (!allPassed) {
    process.exit(1);
  }
}

runCheck();
